action/library/src/main/java/io/knotx/fragments/action/library/cache/CacheAction.java

```diff
@@ -0,0 +1,135 @@
+import io.knotx.fragments.action.library.cache.operations.CacheStore;
+import io.knotx.fragments.action.library.exception.ActionConfigurationException;
+import io.knotx.fragments.api.FragmentContext;
+import io.knotx.fragments.api.FragmentResult;
+import io.knotx.reactivex.fragments.api.FragmentOperation;
+import io.knotx.server.common.placeholders.PlaceholdersResolver;
+import io.reactivex.Maybe;
+import io.reactivex.Single;
+import org.apache.commons.lang3.StringUtils;
+
+public class CacheAction implements SingleAction {
+
+  private final boolean failWhenLookupFails;
+  private final boolean failWhenStoreFails;
+
+  private final ActionLogLevel logLevel;
+  private final String keySchema;
+  private final String alias;
+
+  private final Action doAction;
+
+  private final CacheLookup lookup;
+  private final CacheStore store;
+
+  public static CacheAction create(Cache cache, CacheActionOptions options, String alias,
+      Action doAction) {
+    checkArgument(StringUtils.isBlank(options.getPayloadKey()),
+        () -> new ActionConfigurationException(alias,
+            "CacheAction requires payloadKey value in configuration."));
+    checkArgument(StringUtils.isBlank(options.getCacheKey()),
+        () -> new ActionConfigurationException(alias,
+            "CacheAction requires cacheKey value in configuration."));
+    checkArgument(doAction == null, () -> new ActionConfigurationException(alias,
+        "CacheAction requires doAction configured but none provided"));
+    return new CacheAction(
+        alias,
+        options.getCacheKey(),
+        doAction,
+        ActionLogLevel.fromConfig(options.getLogLevel(), ActionLogLevel.ERROR),
+        new CacheLookup(cache, options.getPayloadKey()),
+        new CacheStore(cache, options.getPayloadKey()),
+        options.isFailWhenLookupFails(),
+        options.isFailWhenStoreFails()
+    );
+  }
+
+  public CacheAction(String alias, String keySchema, Action doAction, ActionLogLevel logLevel,
+      CacheLookup lookup, CacheStore store, boolean failWhenLookupFails,
+      boolean failWhenStoreFails) {
+    this.alias = alias;
+    this.doAction = doAction;
+    this.keySchema = keySchema;
+    this.logLevel = logLevel;
+    this.lookup = lookup;
+    this.store = store;
+    this.failWhenLookupFails = failWhenLookupFails;
+    this.failWhenStoreFails = failWhenStoreFails;
+  }
+
+  @Override
+  public Single<FragmentResult> apply(FragmentContext fragmentContext) {
+    CacheActionLogger logger = CacheActionLogger.create(alias, logLevel);
+    String cacheKey = createCacheKey(fragmentContext);
+
+    return lookup.find(cacheKey, logger)
+        .map(value -> lookup.toResponse(fragmentContext, value))
+        .doOnError(logger::onError)
+        .onErrorResumeNext(this::handleLookupError)
+        .switchIfEmpty(retrieve(fragmentContext, logger)
+            .doOnSuccess(result -> safeSave(logger, cacheKey, result))
```

Comment:
By default we fail when cache store operation throws. But for some customers this behaviour might be acceptable. In either case, we log the exception.